docs/cns.md

```diff
@@ -0,0 +1,110 @@
+# Chain Name Service Draft
+
+The goal of this module is to be a simple registry with the ability to transfer ownership. Auctions and other mechanisms are out of scope and can be implemented alongside or on top of this system (perhaps, in a custodial way).
+
+# State
+
+## Chain name
+
+Chain name contains information about the chain, most importantly, chain ID used to uniquly identify the chain and a list of peer nodes.
+
+Key: `chain-[chainID] -> Chain Name` 
+
+```go
+type ChainName struct {
+	ChainID string
+	Owner sdk.AccAddress
+	Peers []peer
+	SourceURL string
+	SourceCommitHash string
+}
+```
+
+## Offer
+
+Offers are used to transfer ownership of a chain name.
+
+```go
+type Offer struct {
+	ID int32
+	From sdk.AccAddress
+	Amount sdk.Coins
+	Status status // pending | rejected | approved
+}
+```
+
+# Messages
+
+## MsgEditChainName
+
+Chain owner can make edits to the chain name.
+
+```go
+type MsgEditChainName struct {
+	
+}
+```
+
+## MsgTransferChainName
+
+To transfer a name from one owner from another without creating an offer with a zero amount.
+
+```go
+type MsgTransferChainName struct {
+	From string
+	To string
+	ChainID string
+}
+```
+
+## MsgCreateOffer
+
+```go
+type MsgCreateOffer struct {
+	ChainID string
+	Amount sdk.Coins
+}
+```
+
+## MsgRejectOffer
+
+```go
+type MsgRejectOffer struct {
+	ChainID string
+}
+```
+
+## MsgApproveOffer
+
+```go
+type MsgApproveOffer struct {
+	ChainID string
+}
+```
+
+## MsgRemoveOffer
+
+```go
+type MsgRejectOffer struct {
+	ChainID string
+}
+```
+
+## MsgRegisterName
+
+To accuire a name that doesn't have an owner. By default each name has a price. When registering a name, amount equivalent to the price goes to the community pool.
+
+```go
+type MsgRejectOffer struct {
+	ChainID string
+}
+```
+
+# Ideas
+
+- Reserve names for existing chains
+- Names with a chain IDs that have `tesnet-` prefix are registered for free
+- Off-chain tool to check that nodes in peer list are part of a network with a given chain ID. Can loop through all the chain names and create a table.
+- If you're launching a chain and don't want a single entity to be responsible for managing a chain name, use a multisig
+- This should probably be an IBC module
+- Open question: how does relate to IBC, relayers. How can this module be used with IBC.
```

Comment:
ah, began addressing above. Will copy past previous comment here:
> IBC is providing the app hash of all potential chain-ids already. I think the goal should be to leverage the actual connections as the source of updated information.

I think first step should be to replicate what the github is trying to do right now:
 * Chain-ID
 * Seed Nodes
 * Whitelist for administration of those values

I think second step should be some way to associate a chain ID with a list of connection IDs (since there can be multiple connections between the same chain). This is a tricky oracle problem. If we're trusting the "admins" of the chain id, we could extend that to their judgement of which connections are legitimate. There should be a way to confirm a second connection once a first one is made as well but I'm not sure exactly what steps are needed.

For simplicity maybe admin can approve manually is a good first step though, and it's just weak subjectivity as a guarantee?

ApeUnit is working on the flow for a wallet or block explorer to confirm an asset displayed with an account query by utilizing the github + tmcrawl/atlas at the moment. I think this should be the initial goal for the chain-registry as well.